src/control/system/database.go

```diff
@@ -488,7 +488,16 @@ func (db *Database) GroupMap() (*GroupMap, error) {
 
 	gm := newGroupMap(db.data.MapVersion)
 	for _, srv := range db.data.Members.Ranks {
-		if srv.state&AvailableMemberFilter == 0 {
+		// Only members that have been evicted or administratively
+		// excluded should be omitted from the group map. If a member
+		// is actually down, it will be marked dead by swim and moved
+		// into the evicted state eventually.
+		if srv.state&ExcludedMemberFilter != 0 {
+			continue
+		}
+		// Quick sanity-check: Don't include members that somehow have
+		// a nil rank or fabric URI, either.
+		if srv.Rank.Equals(NilRank) || srv.FabricURI == "" {
 			continue
 		}
```

Comment:
When do these happen? Skipping the rank == excluding it. And if we exclude this rank now for map version N, we must deterministically do exactly the same thing for map version N in the future, for instance, even after MS leadership changes. Otherwise, I'd create a new map version that explicitly exclude this rank before distributing the map (seems GroupMap is not a convenient place to do so).

Reply:
By "explicitly exclude" I mean using the state field.

Reply:
> 
> 
> When do these happen? 

Should be never. In fact, after I added this check I tried to write a negative test case for it and realized that it's impossible to apply a raft update with a nil rank because that will trigger the mechanism for selecting/incrementing the next rank. Likewise I'm not sure that a rank could have an empty FabricURI string but it seemed prudent to handle that case.

> Skipping the rank == excluding it. And if we exclude this rank now for map version N, we must deterministically do exactly the same thing for map version N in the future, for instance, even after MS leadership changes. Otherwise, I'd create a new map version that explicitly exclude this rank before distributing the map (seems GroupMap is not a convenient place to do so).

Hmm. As the check was meant to handle supposedly "impossible" scenarios, it seems like it would be better to fail noisily rather than silently skipping over them, then.

Reply:
> Hmm. As the check was meant to handle supposedly "impossible" scenarios, it seems like it would be better to fail noisily rather than silently skipping over them, then.

I'm going to do this in a follow-on, because I'd like to get this landed now to help with some of the test issues we've been seeing, and think a bit more about how to handle this situation. It's an indicator that the system database is corrupted I think, but I'm not certain what should happen in that case. We could shut down this instance but if the corruption somehow got replicated then that won't really help.

Reply:
Exactly. The more defensively I check for unexpected errors, the more difficult (or impossible) the error handling may turn out. Panic'ing is often the only practical handling.